Extending/FileSystemProviders/index.md

```diff
@@ -1,48 +1,159 @@
+:::
+are stored in the format ~/media/
+
+You can set the MediaPathScheme during composition, for example if you wanted to revert back to the V7 methodology in a migrated site:
 ```
+  composition.RegisterUnique<IMediaPathScheme, OriginalMediaPathScheme>();
+```
+
+And you could create your own logic for the path by implementing `IMediaPathScheme`
+
+## Other IFileSystems
+
+Umbraco also registers instances of PhysicalFileSystem for the following parts of Umbraco that persist to 'files':
+
+- MacroPartialsFileSystem
+- PartialViewsFileSystem
+- StylesheetsFileSystem
+- ScriptsFileSystem
+- MvcViewsFileSystem 
 
-Or alternately something like:
+These are accessible via properties on the current registered `IFileSystems` implementation.
 
 ```csharp
-CustomFileSystem fs = FileSystemProviderManager.Current.GetUnderlyingFileSystemProvider("media") as CustomFileSystem;
+IFileSystem macroPartialsFileSystem = Current.FileSystems.MacroPartialsFileSystem;
+```
 
-if (fs != null) {
-    // Do some stuff here
-}
-```   
+or again via dependency injection
+
+```csharp
+  public class FileSystemLocations 
+    {
+        private readonly IFileSystems _fileSystems;
+        public FileSystemLocations(IFileSystems fileSystems)
+        {
+            _fileSystems = fileSystems;
+            var macroPartialsFileSystem = _fileSystems.MacroPartialsFileSystem;
+        }        
+```
 
-Both `IFileSystem` and `FileSystemProviderManager` are located in the `Umbraco.Core.IO` namespace.
+Both `IFileSystem` and `IMediaFileSystem` are located in the `Umbraco.Core.IO` namespace.
 
 ## Custom providers
 
-* [Azure Blob Storage](Azure-Blob-Storage/)
\ No newline at end of file
+There is an Azure Blob Storage provider:
+
+* [Azure Blob Storage](Azure-Blob-Storage/)
+
+### Creating your own FileSystem provider
+Create your filesystem class:
+```csharp
+            public class MyFileSystem : FileSystemWrapper
+            {
+                public MyFileSystem(IFileSystem innerFileSystem)
+                    : base(innerFileSystem)
+                { }
+            }
+```
+The constructor can have more parameters, that will be resolved by the dependency injection container.
+
+Register your new filesystem implementation, in a component:
+```csharp
+  composition.RegisterFileSystem<MyFileSystem>();
+```
+Register the underlying filesystem:
+```csharp
+  composition.RegisterUniqueFor<IFileSystem, MyFileSystem>(...);
+```
+You can inject MyFileSystem wherever it's needed.
+
+Furthermore it's possible to declare a filesystem interface:
+```csharp
+  public interface IMyFileSystem : IFileSystem
+  { }
+```
+Make the class implement the interface, then
+register your filesystem, in a component:
+```csharp
+  composition.RegisterFileSystem<IMyFileSystem, MyFileSystem>();
+  composition.RegisterUniqueFor<IFileSystem, IMyFileSystem>(...);
+```
+You can inject IMyFileSystem wherever it's needed.
+
+### What is Shadowing?
```

Comment:
Is this something you'll run into when working with this? 
Sounds like it's an Umbraco Cloud thing? 🤔 Not entirely sure I understand the context here.. but I might be missing something very obvious 😅

Reply:
I have no idea, I found the explanation in the source code following the adivce on IMyFileSystem... I have no idea if it helps to know it... I reckon we take it out?

Reply:
I have removed it

Reply:
Yes, I agree. It did confuse me a bit, so it might confuse others as well (maybe!) 😅